exporter/metrics.go

```diff
@@ -300,14 +262,22 @@ func (m *metricWithLabels) processRetention(vec deleterMetric) error {
 }
 
 func (m *counterMetric) ProcessMatch(line string, additionalFields map[string]string) (*Match, error) {
-	return m.processMatch(line, func() {
-		m.counter.Inc()
+	return m.processMatch(line, func(value float64) bool {
```

Comment:
This implementation would silently ignore lines where the value evaluates to a number < 0.
It would be better if the callback returned an `error` instead of a `bool`, and if `value < 0` it would return an error saying that values < 0 for counter are not supported and gauge is the correct type for that. If we then return this error from `ProcessMatch` it will be handled in `grok_exporter.go`.

Reply:
I disagree that it would be silent, it would still add to `number_of_lines_ignored`, but I agree that it might be to subtle.

I do not believe that a simple `error` code instead of `bool` will do the trick, as you'll still have a binary return, but instead of your binary options being `ok` and `ignored`, your outcomes will rather be `ok` and `error`.

So I think we should rather take inspiration from the return type of `processMatch` itself and return `bool, error`, with the following matrix defining outcomes:

|             | no error |  error |
|-------------|----------|--------|
| not matched | ignored  | error  |
| matched     | ok       | ok & error      |

This would model the behavior in `grok_exporter.go`, however the `ok & error` output scenario seems rather odd to me. That you can strictly speaking both add to `nErrorsByMetric` and `nMatchesByMetric`.

So I think I'd rather change the matrix to:

|             | no error |  error |
|-------------|----------|--------|
| not matched | ignored  | error  |
| matched     | ok       | error      |

And thus, change `grok_exporter.go` from:
```
128                 match, err := metric.ProcessMatch(line.Line, makeAdditionalFields(line))   
129                 if err != nil {
130                     fmt.Fprintf(os.Stderr, "WARNING: skipping log line: %v\n", err.Error())
131                     fmt.Fprintf(os.Stderr, "%v\n", line.Line)
132                     nErrorsByMetric.WithLabelValues(metric.Name()).Inc()
133                 }
134                 if match != nil {
135                     nMatchesByMetric.WithLabelValues(metric.Name()).Inc()
136                     procTimeMicrosecondsByMetric.WithLabelValues(metric.Name()).Add(float64(time.Since(start).Nanoseconds() / int64(1000)))
137                     matched = true
138                 }
```
to
```
128                 match, err := metric.ProcessMatch(line.Line, makeAdditionalFields(line))   
129                 if err != nil {
130                     fmt.Fprintf(os.Stderr, "WARNING: skipping log line: %v\n", err.Error())
131                     fmt.Fprintf(os.Stderr, "%v\n", line.Line)
132                     nErrorsByMetric.WithLabelValues(metric.Name()).Inc()
133                 }
134                 **else** if match != nil {
135                     nMatchesByMetric.WithLabelValues(metric.Name()).Inc()
136                     procTimeMicrosecondsByMetric.WithLabelValues(metric.Name()).Add(float64(time.Since(start).Nanoseconds() / int64(1000)))
137                     matched = true
138                 }
```
Or is there a specific reason, why we would like to support matching and error at once?

Reply:
That sounds like a good idea, I do not think that it is possible that we get both, a match and an error.

Reply:
Fixed in e3d7bc9